integration/v3_metrics_test.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2016 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package integration
+
+import (
+	"bufio"
+	"io"
+	"log"
+	"net/http"
+	"strconv"
+	"strings"
+	"testing"
+
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"golang.org/x/net/context"
+)
+
+func TestV3ClientMetrics(t *testing.T) {
+	// listen for all prometheus metrics
+	go func() {
+		http.Handle("/metrics", prometheus.Handler())
+		http.ListenAndServe(":27989", nil)
+	}()
+
+	clus := NewClusterV3(t, &ClusterConfig{Size: 3})
+	defer clus.Terminate(t)
+
+	ctxb := context.Background()
+	ctx, cancel := context.WithCancel(ctxb)
+	defer cancel()
+
+	wAPI := toGRPC(clus.RandClient()).Watch
+
+	wStream, err := wAPI.Watch(ctx)
+	if err != nil {
+		t.Fatalf("wAPI.Watch error: %v", err)
+	}
+
+	watchRequest := &pb.WatchRequest{RequestUnion: &pb.WatchRequest_CreateRequest{
+		CreateRequest: &pb.WatchCreateRequest{
+			Key: []byte("foo")}}}
+
+	err = wStream.Send(watchRequest)
+	if err != nil {
+		t.Fatalf("wStream.Send error: %v", err)
+	}
+	watchResponseBefore := sumCountersForMetricAndLabels("grpc_client_msg_received_total", "Watch", "bidi_stream")
+
+	kvAPI := toGRPC(clus.RandClient()).KV
+
+	putTotalBefore := sumCountersForMetricAndLabels("grpc_client_started_total", "Put", "unary")
+	req := &pb.PutRequest{Key: []byte("foo"), Value: []byte("bar")}
+	if _, errp := kvAPI.Put(ctx, req); errp != nil {
+		t.Fatalf("#%d: couldn't put key (%v)", errp)
+	}
+	putTotalAfter := sumCountersForMetricAndLabels("grpc_client_started_total", "Put", "unary")
+
+	if putTotalBefore+1 != putTotalAfter {
+		t.Errorf("grpc_client_started_total should be incremented but recieved : %v\n", putTotalAfter-putTotalBefore)
+	}
+
+	wresp, errw := wStream.Recv()
+	// ensure watcher request created a new watcher
+	if errw != nil {
+		t.Errorf("wStream.Recv error: %v", errw)
+	}
+	if !wresp.Created {
+		t.Errorf("did not create watchid, got %+v", wresp)
+	}
+	if wresp.Canceled {
+		t.Errorf("canceled watcher on create %+v", wresp)
+	}
+
+	watchResponseAfter := sumCountersForMetricAndLabels("grpc_client_msg_received_total", "Watch", "bidi_stream")
+	if watchResponseBefore+1 != watchResponseAfter {
+		t.Errorf("grpc_client_msg_received_total should be incremented but recieved %v\n", watchResponseAfter-watchResponseBefore)
+	}
+}
+
+func sumCountersForMetricAndLabels(metricName string, matchingLabelValues ...string) int {
```

Comment:
Actually i had checked it specific search at this depth is not available for members code it was available to one high level